waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+      (let [service (service-id->service this service-id)
+            delete-result (delete-service api-server-url http-client service)]
+        (comment ;; TODO - remove cached info about this service (once I start caching stuff)
+                 (when delete-result
+                   (remove-failed-instances-for-service! service-id->failed-instances-transient-store service-id)
+                   (scheduler/remove-killed-instances-for-service! service-id)
+                   (swap! service-id->kill-info-store dissoc service-id)))
+        {:result :deleted
+         :message (str "Kubernetes deleted " service-id)})
+      (catch [:status 404] {}
+        (log/warn "[delete-app] Service does not exist:" service-id)
+        {:result :no-such-service-exists
+         :message "Kubernetes reports service does not exist"})
+      (catch [:status 409] e
+        (log/warn "Kubernetes ReplicaSet conflict while deleting"
+                  {:service-id service-id}))))
+
+  (scale-app [this service-id scale-to-instances]
+    (ss/try+
+      (scale-service-to api-server-url http-client
+                     (service-id->service this service-id)
+                     scale-to-instances)
+      {:success true
+       :status 200
+       :result :scaled
+       :message (str "Scaled to " scale-to-instances)}
+      (catch Throwable e
+        {:success false
+         :status 500
+         :result :failed
+         :message (str "Scaling failed: " (.getMessage e))})))
+
+  (retrieve-directory-content [_ service-id instance-id _ relative-directory]
+    ;; TODO - can I implement this?
+    ;; I could just return the log for each pod:
+    ;; /api/v1/namespaces/{ns}/pods/{instance-id}/log
+    ;; I could also do this by starting a little python file server in the home directory.
+    ;; I'd start it with a timeout of 5 minutes or something,
+    ;; and if no requests come in for 5 minutes, then we kill it.
+    [])
+
+  (service-id->state [_ service-id]
+    {:failed-instances (get @service-id->failed-instances-transient-store service-id [])
+     :killed-instances (scheduler/service-id->killed-instances service-id)})
+
+  (state [_]
+    {:service-id->failed-instances @service-id->failed-instances-transient-store
+     :instance-x->failed-instances @instance-x->failed-instances-transient-store}))
+
+(defn kubernetes-scheduler
+  "Returns a new KubernetesScheduler with the provided configuration. Validates the
+  configuration against kubernetes-scheduler-schema and throws if it's not valid."
+  [{:keys [authentication http-options force-kill-after-ms framework-id-ttl
+           service-id->service-description-fn url]}]
+  {:pre [(utils/pos-int? framework-id-ttl)
+         (utils/pos-int? (:conn-timeout http-options))
+         (utils/pos-int? (:socket-timeout http-options))]}
+  (let [http-client (mesos-utils/http-client-factory http-options)]
+    (when authentication
+      (let [refresh-fn (-> authentication :refresh-fn utils/resolve-symbol deref)
```

Comment:
Nitpick: create a separate function for this `let` block.

Reply:
Done.